noir-projects/aztec-nr/aztec/src/state_vars/single_private_mutable.nr

```diff
@@ -0,0 +1,261 @@
+///
+/// Because this private value has no semantic owner, it is up to the application to determine which accounts will
+/// learn of its existence via [crate::note::note_message::NoteMessage::deliver_to].
+///
+/// # Usage
+/// Unlike [crate::state_vars::private_immutable::PrivateImmutable] which is "owned" (requiring wrapping in an
+/// [crate::state_vars::owned::Owned] state variable), SinglePrivateMutable is used directly in storage:
+///
+/// ```noir
+/// #[storage]
+/// struct Storage<Context> {
+///     your_variable: SinglePrivateMutable<YourNote, Context>,
+/// }
+/// ```
+///
+/// Reading from a SinglePrivateMutable nullifies the current note, which restricts the use of this state variable to
+/// situations where race conditions are not a concern. This is commonly the case when dealing with admin-only
+/// functions.
+///    Given that this state variable works with private state it makes sense to use it only when we need the value
+/// to be known by a single individual or a closed set of parties.
+///
+/// # Examples
+///
+/// ## Account contract with signing key rotation
+///
+/// An account contract's signing key can be modeled as a SinglePrivateMutable<PublicKeyNote>. The "current value" of
+/// this state variable holds the active signing key. When the account owner wishes to update the signing key, they
+/// invoke the `replace` function, providing a new public key note as input.
+///
+/// ## Private Token
+///
+/// A private token's admin and total supply can be stored in two single private mutable state variables:
+/// ```noir
+/// #[storage]
+/// struct Storage<Context> {
+///     admin: SinglePrivateMutable<AddressNote, Context>,
+///     total_supply: SinglePrivateMutable<UintNote, Context>,
+///     balances: Owned<BalanceSet<Context>, Context>,
+/// }
+/// ```
+///
+/// When the mint or burn functions are invoked, we check that the message sender corresponds to the current value of
+/// the admin state variable. Then we update the total supply and balances. The resulting note message can then be
+/// delivered to an account whose encryption keys are known to a closed set of parties. With this approach, only
+/// the admin can modify the total supply of the token, but anyone with the relevant encryption keys can audit it.
+///
+/// # Privacy
+///
+/// Initialization of this state variable leaks information because the initialization nullifier can be predicted by
+/// anyone with the knowledge of the contract address (nullifiers are being hashed with the contract address by the
+/// kernel circuits before being inserted into the nullifier tree and the storage slot can be guessed/brute forced).
+/// This problem is generally avoided by including owner's nullifying secret in the nullifier preimage. Here this is
+/// not a possibility because there is no owner.
+///
+/// This issue is shared with contract initialization nullifiers, and is tracked by
+/// https://github.com/AztecProtocol/aztec-packages/issues/17128
```

Comment:
I think it's quite likely that this state var will also prove desirable to be used in "closed set" settings where there could be a notion of shared keys which might make it possible to use the nsk of those shared keys in the initialization nullifier.

But would wait with tackling this as we don't have enough usecases yet so it all feels a bit abstract now.

(In the `SinglePrivateImmutable` I use the contract's `nsk`)